tiny_dnn/util/serialization_layer_list.h

```diff
@@ -1,25 +1,32 @@
 // Copyright (c) 2017, Taiga Nomi
-#ifndef CNN_NO_SERIALIZATION
+#pragma once
 
-CNN_REGISTER_LAYER(average_pooling_layer, avepool);
-CNN_REGISTER_LAYER(batch_normalization_layer, batchnorm);
-CNN_REGISTER_LAYER(concat_layer, concat);
-CNN_REGISTER_LAYER(convolutional_layer, conv);
-CNN_REGISTER_LAYER(dropout_layer, dropout);
-CNN_REGISTER_LAYER(fully_connected_layer, fully_connected);
-CNN_REGISTER_LAYER(linear_layer, linear);
-CNN_REGISTER_LAYER(lrn_layer, lrn);
-CNN_REGISTER_LAYER(max_pooling_layer, maxpool);
-CNN_REGISTER_LAYER(power_layer, power);
-CNN_REGISTER_LAYER(slice_layer, slice);
-CNN_REGISTER_LAYER(elementwise_add_layer, elementwise_add);
+#include "tiny_dnn/layers/layers.h"
 
-CNN_REGISTER_LAYER(sigmoid_layer, sigmoid);
-CNN_REGISTER_LAYER(tanh_layer, tanh);
-CNN_REGISTER_LAYER(relu_layer, relu);
-CNN_REGISTER_LAYER(softmax_layer, softmax);
-CNN_REGISTER_LAYER(leaky_relu_layer, leaky_relu);
-CNN_REGISTER_LAYER(elu_layer, elu);
-CNN_REGISTER_LAYER(tanh_p1m2_layer, tanh_scaled);
+namespace tiny_dnn {
 
-#endif
+template <typename T>
+void register_layers(T* h) {
+  h->template register_layer<average_pooling_layer>("avepool");
```

Comment:
Yes, it is necessary with gcc. Microsoft's VC++ doesn't need it though.
I guess the relationship is like Mozilla Firefox and Internet Explorer.